libraries/AP_Compass/AP_Compass_RM3100.cpp

```diff
@@ -0,0 +1,208 @@
+        return false;
+    }
+
+    // high retries for init
+    dev->set_retries(10);
+
+    // maybe add a test for revision id value here and goto fail if not correct
+    // it is not clear if revision id is the same for all compasses, in the same manner as a whoami
+
+    dev->setup_checked_registers(8);
+
+    dev->write_register(RM3100_TMRC_REG, TMRC, true); // cycle count z
+    dev->write_register(RM3100_CMM_REG, CMM, false); // CMM configuration
+    dev->write_register(RM3100_CCX1_REG, CCP1, true); // cycle count x
+    dev->write_register(RM3100_CCX0_REG, CCP0, true); // cycle count x
+    dev->write_register(RM3100_CCY1_REG, CCP1, true); // cycle count y
+    dev->write_register(RM3100_CCY0_REG, CCP0, true); // cycle count y
+    dev->write_register(RM3100_CCZ1_REG, CCP1, true); // cycle count z
+    dev->write_register(RM3100_CCZ0_REG, CCP0, true); // cycle count z
+
+    _scaler = (1 / GAIN_CC200) * UTESLA_TO_MGAUSS; // has to be changed if using a different cycle count
+
+    // lower retries for run
+    dev->set_retries(3);
+    
+    dev->get_semaphore()->give();
+
+    /* register the compass instance in the frontend */
+    compass_instance = register_compass();
+
+    printf("Found a RM3100 at address 0x%x as compass %u\n", dev->get_bus_address(), compass_instance);
+    
+    set_rotation(compass_instance, rotation);
+
+    if (force_external) {
+        set_external(compass_instance, true);
+    }
+    
+    dev->set_device_type(DEVTYPE_RM3100);
+    set_dev_id(compass_instance, dev->get_bus_id());
+
+    // call timer() at 80Hz
+    dev->register_periodic_callback(1000000U/80U,
+                                    FUNCTOR_BIND_MEMBER(&AP_Compass_RM3100::timer, void));
+
+    return true;
+
+//fail:
+//    dev->get_semaphore()->give();
+//    return false;
+}
+
+void AP_Compass_RM3100::timer()
+{
+    struct PACKED {
+        uint8_t magx_2;
+        uint8_t magx_1;
+        uint8_t magx_0;
+        uint8_t magy_2;
+        uint8_t magy_1;
+        uint8_t magy_0;
+        uint8_t magz_2;
+        uint8_t magz_1;
+        uint8_t magz_0;
+    } data;
+    Vector3f field;
+
+    uint32_t magx = 0;
+    uint32_t magy = 0;
+    uint32_t magz = 0;
+
+    // check data ready on 3 axis
+    uint8_t status;
+    if (!dev->read_registers(RM3100_STATUS_REG, (uint8_t *)&status, 1)) {
+        goto check_registers;
+    }
+
+    if (!(status & 0x80)) {
+        // data not available yet
+        goto check_registers;
+    }
+
+    if (!dev->read_registers(RM3100_MX2_REG, (uint8_t *)&data, sizeof(data))) {
+        goto check_registers;
+    }
+
+
+
+    magx = ((uint32_t)data.magx_2 << 16) | ((uint32_t)data.magx_1 << 8) | (uint32_t)data.magx_0;
+    magy = ((uint32_t)data.magy_2 << 16) | ((uint32_t)data.magy_1 << 8) | (uint32_t)data.magy_0;
+    magz = ((uint32_t)data.magz_2 << 16) | ((uint32_t)data.magz_1 << 8) | (uint32_t)data.magz_0;
+
+    if (magx & 0x800000) {
```

Comment:
this code at least deserves a comment. I think what you are doing is extracting 24 bit 2s complement signed values. Why not take the 32 bit values and then shift left by 8 bits to make a signed 32 bit number?

Reply:
Seems better yes ! Then divide by 2^8 right ?

Reply:
I simplified the syntax as suggested and flight-tested it.